### PR TITLE
Feature/dsm gained

### LINKED ISSUE #5: Add the rewards calculation and visualization
## Description
Currently only the challenges status is displayed. We should add also the amount of DSM that each user has gained. 

The rewards should be computed based on how much each single challenge is worth. The values are the following. 

### Primer Phase 1
- 200 DSM per each accepted referral
- 20 DSM for the post
- 10 DSM for the like 
- 50 DSM if someone referred the user and the referral was accepted

### Primer Phase 2 
- 30 DSM for the reaction
- 300 DSM for the validator creation

### Primer Phase 3 
- 100 DSM for the poll
- 10 DSM for the poll answer
- 50 DSM for the multimedia post

Plus, the following rewards should be given based on the number of blocks validated: 

| Pre-commits signed | Tokens rewarded | 
| :----------------: | :-------------: |
| 1 | 1 |
| 4 | 2 |
| 18 | 3 |
| 78 | 7 |
| 331 | 12 |
| 1,416 | 21 |
| 6,044 | 40 |
| 25,796 | 73 |
| 110,097 | 135 |
| 469,894 | 250 |

### Primer Phase 4 
- 50 DSM for the account
- 50 DSM for the reaction
- 50 DSM for the update 

Plus, the following rewards should be given based on the number of blocks validated: 

| Pre-commits signed | Tokens rewarded | 
| :----------------: | :-------------: |
| 345,600 | 1,500 |
| 353,480 | 1,750 |
| 361,540 | 2,040 |
| 369,790 | 2,380 |
| 378,220 | 2,780 |
| 386,850 | 3,240 |
| 395,670 | 3,780 |
| 404,690 | 4,410 |
| 413,920 | 5,140 |
| 423,360 | 6,000 |

### Primer Phase 5 
- 25 DSM for the hashtag
- 50 DSM for the account 
- 25 DSM for the report
- 50 DSM for the tag 
- 50 DSM for the update 

Plus, the following rewards should be given based on the number of blocks validated: 

| Pre-commits signed | Tokens rewarded | 
| :----------------: | :-------------: |
| 476,534 | 1,500 |
| 488,497 | 1,750 |
| 500,760 | 2,040 |
| 513,331 | 2,380 |
| 526,217 | 2,780 |
| 539,427 | 3,240 |
| 552,969 | 3,780 |
| 566,851 | 4,410 |
| 581,081 | 5,140 |
| 595,668 | 6,000 |   

### Validating summer
Each validator that has signed at least 462,631 precommits should receive 6,000 DSM as a reward.

### PR DESCRIPTION
**Pull request by ryuash**
## Description

closes #5 

updates:
phase 1:
- added is referred and successful referred count. dsm gained includes new data
Screen Shot 2020-10-07 at 1 22 45 PM

-users page
- not all users are connected to desmos primer phase so i had to improvise
Screen Shot 2020-10-07 at 1 23 05 PM

- ill highlight some code i need you to review (the calculations make sure it's correct)
- ill make an issue to refactor some of the display code in the future (it's a lil brute forced right now)
## Checklist
- [x] Ran Linting
- [x] Targeted PR against correct branch.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote unit tests.
- [ ] Added an entry to the `CHANGELOG.md` file.
- [x] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: src/utils/desmos_primer_dsm_calculate.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  else if (precommit >= 25796) total = 73;
+  else if (precommit >= 6044) total = 40;
+  else if (precommit >= 1416) total = 21;
+  else if (precommit >= 331) total = 12;
+  else if (precommit >= 78) total = 7;
+  else if (precommit >= 18) total = 3;
+  else if (precommit >= 4) total = 2;
+  else if (precommit >= 1) total = 1;
+  return total;
+};
```
Comment (ryuash) [inline src/utils/desmos_primer_dsm_calculate.ts:23]: review please @RiccardoM

File: src/utils/desmos_primer_dsm_calculate.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  else if (precommit >= 404690) total = 4410;
+  else if (precommit >= 395670) total = 3780;
+  else if (precommit >= 386850) total = 3240;
+  else if (precommit >= 378220) total = 2780;
+  else if (precommit >= 369790) total = 2380;
+  else if (precommit >= 361540) total = 2040;
+  else if (precommit >= 353480) total = 1750;
+  else if (precommit >= 345600) total = 1500;
+  return total;
+};
```
Comment (ryuash) [inline src/utils/desmos_primer_dsm_calculate.ts:38]: @RiccardoM

File: src/utils/desmos_primer_dsm_calculate.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  else if (precommit >= 566851) total = 4410;
+  else if (precommit >= 552969) total = 3780;
+  else if (precommit >= 539427) total = 3240;
+  else if (precommit >= 526217) total = 2780;
+  else if (precommit >= 513331) total = 2380;
+  else if (precommit >= 500760) total = 2040;
+  else if (precommit >= 488497) total = 1750;
+  else if (precommit >= 476534) total = 1500;
+  return total;
+};
```
Comment (ryuash) [inline src/utils/desmos_primer_dsm_calculate.ts:53]: @RiccardoM

File: src/utils/desmos_primer_dsm_calculate.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+ * Used to get total DSM gained for phase one
+ */
+export const getPhaseOneDSM = (data:DesmosPrimerPhaseOne) => {
+  let total = 0;
+  if (data.createdPost) total += 20;
+  if (data.likedPost) total += 10;
+  if (data.isReferred) total += 50;
+  if (data.successReferredCount) total += (data.successReferredCount * 200);
+  return total;
+};
```
Comment (ryuash) [inline src/utils/desmos_primer_dsm_calculate.ts:69]: @RiccardoM

File: src/utils/desmos_primer_dsm_calculate.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+// 300 DSM for the validator creation
+/**
+ * Used to get total DSM gained for phase two
+ */
+export const getPhaseTwoDSM = (data:DesmosPrimerPhaseTwo) => {
+  let total = 0;
+  if (data.addedReaction) total += 30;
+  if (data.createdValidator) total += 300;
+  return total;
+};
```
Comment (ryuash) [inline src/utils/desmos_primer_dsm_calculate.ts:81]: @RiccardoM

File: src/utils/desmos_primer_dsm_calculate.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+ * Used to get total DSM gained for phase three
+ */
+export const getPhaseThreeDSM = (data:DesmosPrimerPhaseThree) => {
+  let total = 0;
+  if (data.createdPoll) total += 100;
+  if (data.answeredPoll) total += 10;
+  if (data.createdMultimediaPost) total += 50;
+  total += getPhaseThreePrecommitsDSM(data.precommitsSigned);
+  return total;
+};
```
Comment (ryuash) [inline src/utils/desmos_primer_dsm_calculate.ts:96]: @RiccardoM

File: src/utils/desmos_primer_dsm_calculate.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+ * Used to get total DSM gained for phase four
+ */
+export const getPhaseFourDSM = (data:DesmosPrimerPhaseFour) => {
+  let total = 0;
+  if (data.validatorAddress) total += 50;
+  if (data.createdAccount) total += 50;
+  if (data.registeredReaction) total += 50;
+  total += getPhaseFourPrecommitsDSM(data.precommitsSigned);
+  return total;
+};
```
Comment (ryuash) [inline src/utils/desmos_primer_dsm_calculate.ts:111]: @RiccardoM

File: src/utils/desmos_primer_dsm_calculate.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+export const getPhaseFiveDSM = (data:DesmosPrimerPhaseFive) => {
+  let total = 0;
+  if (data.createdPostWithHashtag) total += 25;
+  if (data.createdReport) total += 20;
+  if (data.createdPostWithTag) total += 50;
+  if (data.createdOrUpdatedProfile) total += 50;
+  if (data.validatorAddress) total += 50;
+  total += getPhaseFivePrecommitsDSM(data.precommitsSigned);
+  return total;
+};
```
Comment (ryuash) [inline src/utils/desmos_primer_dsm_calculate.ts:131]: @RiccardoM

File: src/utils/desmos_primer_dsm_calculate.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Used to get total DSM gained for phase one referrals
+ */
+export const getReferralDSM = (data:DesmosPrimerReferral) => {
+  let total = 0;
+  total = data.referrals?.filter((x) => x.accepted).length * 200;
+  return total;
+};
```
Comment (ryuash) [inline src/utils/desmos_primer_dsm_calculate.ts:140]: @RiccardoM

File: src/utils/desmos_validating_summer_dsm_calculate.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Used to get total DSM gained for validating summer
+ */
+export const getValidatingSummerDSM = (data:any) => {
+  return data.precommitsSigned >= 462631 ? 6000 : 0;
+};
```
Comment (ryuash) [inline src/utils/desmos_validating_summer_dsm_calculate.ts:6]: @RiccardoM

File: src/utils/desmos_primer_dsm_calculate.ts (hunk 8 above)
Comment (RiccardoM) [reply src/utils/desmos_primer_dsm_calculate.ts:131]: Should be 25 for report

**Review by RiccardoM — COMMENTED** [review]
Looks good to me. Can we apply some changes @ryuash? 

1. Remove the rewards things from the Phase 1 tab. Since they're already present inside the specific tab, I wouldn't duplicate this. 

2. Can we add a row to the bottom of each tab, telling how many total DSM were given out inside each Primer Phase / Validating Summer / to each user?

**Review by RiccardoM — APPROVED** [review]

**Pull request merged**